File: coding_skill_cpp/17_raii_pattern.cpp
/* C++ RAII - 소멸자 자동 호출 */
#include <iostream>
#include <fstream>
using namespace std;

class FileResource {
    FILE* file;
public:
    FileResource(const char* filename) {
        file = fopen(filename, "w");
        cout << "[RAII] 파일 열림" << endl;
    }
    ~FileResource() {
        if (file) fclose(file);
        cout << "[RAII] 파일 자동 닫힘" << endl;
    }
};

int main() {
    cout << "=== C++ RAII ===" << endl;
    {
        FileResource res("test.txt");
    }  // 자동 소멸
    cout << "블록 종료!" << endl;
    return 0;
}

File: coding_skill_py/17_raii_pattern.py
"""Python RAII - with 문"""

class FileResource:
    def __init__(self, filename: str):
        self.filename = filename
        self.file = None
    
    def __enter__(self):
        self.file = open(self.filename, 'w')
        print("[RAII] 파일 열림")
        return self.file
    
    def __exit__(self, *args):
        if self.file:
            self.file.close()
        print("[RAII] 파일 자동 닫힘")

def main():
    print("=== Python RAII (with) ===")
    with FileResource("test.txt") as f:
        f.write("Hello")
    print("블록 종료!")

if __name__ == "__main__":
    main()
